sphinx/ext/intersphinx.py

```diff
@@ -233,11 +234,31 @@ def fetch_inventory_group(
                     failures.append(err.args)
                     continue
                 if invdata:
+                    old_uris = set()
+                    # Find the URIs of the *existing* projects that are to
+                    # be removed because there are being overwritten.
+                    #
+                    # If the current cache contains some (project, uri) pair
+                    # say ("foo", "foo.com") and if the new intersphinx dict
+                    # contains the pair ("bar", "foo.com"), we need to remove
+                    # the first entry and use the second one.
+                    for old_uri, (old_name, _expiry, _data) in cache.items():
+                        if old_name == name:
+                            old_uris.add(old_uri)
+
+                    for old_uri in old_uris:
+                        del cache[old_uri]
+
+                    if uri in cache:
+                        # ensure that the cache data is moved to the end
+                        # when setting `cache[uri] = ...`
+                        del cache[uri]
```

Comment:
I interpret the `old_uris` set as a collection of URIs to expire from the cache, and that makes sense.

What I'm wondering about is whether there's risk about concurrent access to the `cache` object, given that `fetch_inventory_group` is [invoked from a thread pool](https://github.com/picnixz/sphinx/blob/8bbcd83d76afd40a9db983085073926ff5a14c37/sphinx/ext/intersphinx.py#L280-L289).  I think those threads will reawaken somewhat unpredictably based on the response times for each fetched inventory.

Had you considered clearing expired cache entries before the fetches begin?  (I guess this way avoids duplicate code to check cache expiry times)

Reply:
I did not consider that. Good catch for testing whether this can cause an issue or not. I'll think about it now.

Reply:
There might indeed be a race condition where two threads remove a cached entry at the same time. Either I use `.pop(..., None)` to ensure that the entries are not deleted twice by two different threads, or I can add a simple lock (though I prefer not to since it's quite consuming). 

As for moving the cache data at the end, this should be fine and if two threads update the inventory data at the same time, then I don't know which one should be given the priority. So, even with a lock, some inventory may be incorrectly updated.